lib/util/propTypes.js

```diff
@@ -0,0 +1,671 @@
+    const sourceCode = context.getSourceCode();
+    const propWrapperFunctions = new Set(context.settings.propWrapperFunctions || []);
+
+    /**
+     * Helper for accessing the current scope in the stack.
+     * @param {string} key The name of the identifier to access. If omitted, returns the full scope.
+     * @param {ASTNode} value If provided sets the new value for the identifier.
+     * @returns {Object|ASTNode} Either the whole scope or the ASTNode associated with the given identifier.
+     */
+    function typeScope(key, value) {
+      if (arguments.length === 0) {
+        return stack[stack.length - 1];
+      } else if (arguments.length === 1) {
+        return stack[stack.length - 1][key];
+      }
+      stack[stack.length - 1][key] = value;
+      return value;
+    }
+
+    /**
+     * Checks if prop should be validated by plugin-react-proptypes
+     * @param {String} validator Name of validator to check.
+     * @returns {Boolean} True if validator should be checked by custom validator.
+     */
+    function hasCustomValidator(validator) {
+      return customValidators.indexOf(validator) !== -1;
+    }
+
+    /**
+     * Resolve the type annotation for a given node.
+     * Flow annotations are sometimes wrapped in outer `TypeAnnotation`
+     * and `NullableTypeAnnotation` nodes which obscure the annotation we're
+     * interested in.
+     * This method also resolves type aliases where possible.
+     *
+     * @param {ASTNode} node The annotation or a node containing the type annotation.
+     * @returns {ASTNode} The resolved type annotation for the node.
+     */
+    function resolveTypeAnnotation(node) {
+      let annotation = node.typeAnnotation || node;
+      while (annotation && (annotation.type === 'TypeAnnotation' || annotation.type === 'NullableTypeAnnotation')) {
+        annotation = annotation.typeAnnotation;
+      }
+      if (annotation.type === 'GenericTypeAnnotation' && typeScope(annotation.id.name)) {
+        return typeScope(annotation.id.name);
+      }
+      return annotation;
+    }
+
+    /**
+     * Creates the representation of the React props type annotation for the component.
+     * The representation is used to verify nested used properties.
+     * @param {ASTNode} annotation Type annotation for the props class property.
+     * @return {Object} The representation of the declaration, empty object means
+     *    the property is declared without the need for further analysis.
+     */
+    function buildTypeAnnotationDeclarationTypes(annotation, parentName, seen) {
+      if (typeof seen === 'undefined') {
+        // Keeps track of annotations we've already seen to
+        // prevent problems with recursive types.
+        seen = new Set();
+      }
+      if (seen.has(annotation)) {
+        // This must be a recursive type annotation, so just accept anything.
+        return {};
+      }
+      seen.add(annotation);
+
+      switch (annotation.type) {
```

Comment:
not a huge deal, but it'd be great to avoid a switch statement

Reply:
hmm do you mean a map with handlers for each type? there's also a bunch of other huge switches in here (like in other parts of the codebase) which could also get the same treatment

Reply:
if/else, or a map, yes

Reply:
Not sure how 4-5 chained else-ifs are better than a switch? anyway will try out a map

Reply:
they don't suffer from unintentional fallthrough.